posthog/special_migrations/utils.py

```diff
@@ -0,0 +1,150 @@
+from datetime import datetime
+from typing import Optional
+
+from django.db import transaction
+
+from posthog.celery import app
+from posthog.constants import AnalyticsDBMS
+from posthog.models.special_migration import MigrationStatus, SpecialMigration
+from posthog.special_migrations.setup import DEPENDENCY_TO_SPECIAL_MIGRATION
+
+
+def execute_op(database: AnalyticsDBMS, sql: str, timeout_seconds: int, query_id: str):
+    if database == AnalyticsDBMS.CLICKHOUSE:
+        execute_op_clickhouse(sql, query_id, timeout_seconds)
+        return
+
+    execute_op_postgres(sql, query_id)
+
+
+def execute_op_clickhouse(sql: str, query_id: str, timeout_seconds: int):
+    from ee.clickhouse.client import sync_execute
+
+    sync_execute(f"/* {query_id} */ " + sql, settings={"max_execution_time": timeout_seconds})
+
+
+def execute_op_postgres(sql: str, query_id: str):
+    from django.db import connection
+
+    with connection.cursor() as cursor:
+        cursor.execute(f"/* {query_id} */ " + sql)
+
+
+def process_error(migration_instance: SpecialMigration, error: Optional[str]):
+    update_special_migration(
+        migration_instance=migration_instance,
+        status=MigrationStatus.Errored,
+        last_error=error or "",
+        finished_at=datetime.now(),
+        lock_row=True,
+    )
+
+    from posthog.special_migrations.runner import attempt_migration_rollback
+
+    attempt_migration_rollback(migration_instance)
+
+
+def trigger_migration(migration_instance: SpecialMigration, fresh_start=True):
+    from posthog.tasks.special_migrations import run_special_migration
+
+    task = run_special_migration.delay(migration_instance.name, fresh_start)
+
+    update_special_migration(
+        migration_instance=migration_instance, celery_task_id=str(task.id),
+    )
+
+
+def force_stop_migration(migration_instance: SpecialMigration, error: str = "Force stopped by user"):
+    """
+    In theory this is dangerous, as it can cause another task to be lost 
+    `revoke` with `terminate=True` kills the process that's working on the task
+    and there's no guarantee the task will not already be done by the time this happens.
+    See: https://docs.celeryproject.org/en/stable/reference/celery.app.control.html#celery.app.control.Control.revoke
+    However, this is generally ok for us because:
+    1. Given these are long-running migrations, it is statistically unlikely it will complete during in between 
+    this call and the time the process is killed
+    2. Our Celery tasks are not essential for the functioning of PostHog, meaning losing a task is not the end of the world
+    """
+
+    app.control.revoke(migration_instance.celery_task_id, terminate=True)
+    process_error(migration_instance, error)
+
+
+def force_rollback_migration(migration_instance: SpecialMigration):
```

Comment:
do we need this, why not just call `attempt_migration_rollback` directly?

Reply:
probably can - thinking back I probably split this out because the import might be confusing in the middle of the body of a larger func

Reply:
100% This seems a bit bloaty

Reply:
will change in the API PR which calls this code